File: backend/src/routes/opportunities.js
import express from "express";
import { recommendationService } from "../services/recommendationService.js";

const router = express.Router();

router.get("/", async (req, res) => {
  try {
    const { type, skill, difficulty, max_hours } = req.query;
    const student = {
      goal: "AI Engineer",
      available_hours_per_week: 6.0,
      skills: [{ name: "Python", level: "Intermediate" }],
      interests: ["AI", "Machine Learning"],
    };

    const allOpps = await recommendationService.getRecommendations(student);

    const filtered = allOpps.filter((opp) => {
      if (type && !opp.type.toLowerCase().includes(type.toLowerCase())) {
        return false;
      }
      if (
        skill &&
        !opp.skills_developed.some((s) => s.toLowerCase().includes(skill.toLowerCase()))
      ) {
        return false;
      }
      if (difficulty && !opp.difficulty.toLowerCase().includes(difficulty.toLowerCase())) {
        return false;
      }
      if (max_hours && opp.hours_per_week > parseFloat(max_hours)) {
        return false;
      }
      return true;
    });

    return res.json(filtered);
  } catch (err) {
    console.error("Opportunities list error:", err);
    return res.status(500).json({ error: "Failed to fetch opportunities" });
  }
});

router.get("/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const student = { goal: "AI Engineer", available_hours_per_week: 6.0 };
    const allOpps = await recommendationService.getRecommendations(student);
    const found = allOpps.find((opp) => opp.id.toLowerCase() === id.toLowerCase());

    if (!found) {
      return res.status(404).json({ error: `Opportunity '${id}' not found` });
    }

    return res.json(found);
  } catch (err) {
    console.error("Opportunity detail error:", err);
    return res.status(500).json({ error: "Failed to fetch opportunity detail" });
  }
});

export default router;

File: backend/app/routes/opportunities.py
from typing import Optional, List
from fastapi import APIRouter, Query, HTTPException
from app.models.student import StudentProfile
from app.models.opportunity import OpportunityItem
from app.services.recommendation_service import recommendation_service

router = APIRouter(prefix="/api/opportunities", tags=["opportunities"])

@router.get("", response_model=List[OpportunityItem])
async def list_opportunities(
    type: Optional[str] = Query(None),
    skill: Optional[str] = Query(None),
    difficulty: Optional[str] = Query(None),
    max_hours: Optional[float] = Query(None)
):
    student = StudentProfile()
    all_opps = await recommendation_service.get_recommendations(student)

    filtered = []
    for opp in all_opps:
        if type and type.lower() not in opp.type.lower():
            continue
        if skill and not any(skill.lower() in s.lower() for s in opp.skills_developed):
            continue
        if difficulty and difficulty.lower() not in opp.difficulty.lower():
            continue
        if max_hours and opp.hours_per_week > max_hours:
            continue
        filtered.append(opp)

    return filtered

@router.get("/{id}", response_model=OpportunityItem)
async def get_opportunity_detail(id: str):
    student = StudentProfile()
    all_opps = await recommendation_service.get_recommendations(student)
    for opp in all_opps:
        if opp.id.lower() == id.lower():
            return opp
    raise HTTPException(status_code=404, detail=f"Opportunity '{id}' not found")
